diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index 0023abf..c5b5df9 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -32,6 +32,7 @@
 #include <stdlib.h>
 #include <stdarg.h>
 #include <stdio.h>
+#include <string.h>
 #include <sys/types.h>
 #include <sys/mman.h>
 #include <unistd.h>
@@ -547,6 +548,22 @@ extern void __wine_syscall_dispatcher_instrumentation(void);
 static void *instrumentation_callback;
 static pthread_mutex_t instrumentation_callback_mutex = PTHREAD_MUTEX_INITIALIZER;
 
+// Denuvo bypass controlled by WINE_DENUVO=1
+static int denuvo_enabled = 0;
+static void *TargetSysHandler = 0;
+static unsigned int spoof_leaf1_eax = 0;
+static unsigned int spoof_leaf1_ebx = 0;
+static unsigned int spoof_leaf1_ecx = 0;
+static unsigned int spoof_leaf1_edx = 0;
+static unsigned int spoof_leaf40000000_eax = 0;
+static unsigned int spoof_leaf40000000_ebx = 0;
+static unsigned int spoof_leaf40000000_ecx = 0;
+static unsigned int spoof_leaf40000000_edx = 0;
+static unsigned int spoof_leaf40000001_eax = 0;
+static unsigned int spoof_leaf40000001_ebx = 0;
+static unsigned int spoof_leaf40000001_ecx = 0;
+static unsigned int spoof_leaf40000001_edx = 0;
+
 void set_process_instrumentation_callback( void *callback )
 {
     void *ptr = (char *)user_shared_data + page_size;
@@ -2283,6 +2300,103 @@ static void segv_handler( int signal, siginfo_t *siginfo, void *_sigcontext )
 
     save_context( data, &context, sigcontext );
 
+    if (denuvo_enabled)
+    {
+        unsigned int leaf;
+        unsigned int subleaf;
+        ucontext_t *uc = sigcontext;
+        unsigned char *rip = (unsigned char *)uc->uc_mcontext.gregs[REG_RIP];
+
+        leaf = uc->uc_mcontext.gregs[REG_RAX];
+        subleaf = uc->uc_mcontext.gregs[REG_RCX];
+        if ((siginfo->si_code == SI_KERNEL || leaf == 0x336933) && rip[0] == 0x0F && rip[1] == 0xA2) {
+            switch (leaf) {
+                case 1:
+                    uc->uc_mcontext.gregs[REG_RAX] = spoof_leaf1_eax;
+                    uc->uc_mcontext.gregs[REG_RBX] = spoof_leaf1_ebx;
+                    uc->uc_mcontext.gregs[REG_RCX] = spoof_leaf1_ecx | (TargetSysHandler ? 0 : (0x1 << 31));
+                    uc->uc_mcontext.gregs[REG_RDX] = spoof_leaf1_edx;
+                    break;
+
+                case 0x40000000:
+                    uc->uc_mcontext.gregs[REG_RAX] = spoof_leaf40000000_eax;
+                    uc->uc_mcontext.gregs[REG_RBX] = spoof_leaf40000000_ebx;
+                    uc->uc_mcontext.gregs[REG_RCX] = spoof_leaf40000000_ecx;
+                    uc->uc_mcontext.gregs[REG_RDX] = spoof_leaf40000000_edx;
+                    break;
+
+                case 0x40000001:
+                    uc->uc_mcontext.gregs[REG_RAX] = spoof_leaf40000001_eax;
+                    uc->uc_mcontext.gregs[REG_RBX] = spoof_leaf40000001_ebx;
+                    uc->uc_mcontext.gregs[REG_RCX] = spoof_leaf40000001_ecx;
+                    uc->uc_mcontext.gregs[REG_RDX] = spoof_leaf40000001_edx;
+                    break;
+
+                case 0x80000002:
+                    uc->uc_mcontext.gregs[REG_RAX] = 0x756E6544;
+                    uc->uc_mcontext.gregs[REG_RBX] = 0x4F774F76;
+                    uc->uc_mcontext.gregs[REG_RCX] = 0x55504320;
+                    uc->uc_mcontext.gregs[REG_RDX] = 0x31204020;
+                    break;
+
+                case 0x80000003:
+                    uc->uc_mcontext.gregs[REG_RAX] = 0x20373333;
+                    uc->uc_mcontext.gregs[REG_RBX] = 0x007A4847;
+                    uc->uc_mcontext.gregs[REG_RCX] = 0x00000000;
+                    uc->uc_mcontext.gregs[REG_RDX] = 0x00000000;
+                    break;
+
+                case 0x80000004:
+                    uc->uc_mcontext.gregs[REG_RAX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RBX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RCX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RDX] = 0x0;
+                    break;
+
+                case 0x336933:
+                    MESSAGE("Spoofing CPUID leaf %x\n", leaf);
+                    TargetSysHandler = (void *)uc->uc_mcontext.gregs[REG_RCX];
+                    patch_kuser_shared_data();
+                    uc->uc_mcontext.gregs[REG_RAX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RBX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RCX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RDX] = 0x0;
+                    break;
+
+                case 0x336967:
+                    MESSAGE("Setting Faketime to %llx... \n", (long long)uc->uc_mcontext.gregs[REG_RCX]);
+                    SERVER_START_REQ( set_faketime )
+                    {
+                        req->faketime = uc->uc_mcontext.gregs[REG_RCX];
+                        wine_server_call( req );
+                    }
+                    SERVER_END_REQ;
+                    uc->uc_mcontext.gregs[REG_RAX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RBX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RCX] = 0x0;
+                    uc->uc_mcontext.gregs[REG_RDX] = 0x0;
+                    break;
+
+                default:
+                    syscall(SYS_arch_prctl, ARCH_SET_CPUID, 1);
+                    __asm__ volatile(
+                            "cpuid"
+                            : "=a"(uc->uc_mcontext.gregs[REG_RAX]),
+                              "=b"(uc->uc_mcontext.gregs[REG_RBX]),
+                              "=c"(uc->uc_mcontext.gregs[REG_RCX]),
+                              "=d"(uc->uc_mcontext.gregs[REG_RDX])
+                            : "a"(leaf), "c"(subleaf)
+                            : "memory"
+                        );
+                    syscall(SYS_arch_prctl, ARCH_SET_CPUID, 0);
+            }
+
+            uc->uc_mcontext.gregs[REG_RIP] += 2;
+            leave_handler( data, sigcontext );
+            return;
+        }
+    }
+
     switch(TRAP_sig(sigcontext))
     {
     case TRAP_x86_OFLOW:   /* Overflow exception */
@@ -2616,6 +2730,24 @@ static void sigsys_handler( int signal, siginfo_t *siginfo, void *_sigcontext )
     }
 #endif
 
+    if (denuvo_enabled)
+    {
+        ucontext_t *ctx = sigcontext;
+
+        __uint128_t *xmm_regs = (__uint128_t *)ctx->uc_mcontext.fpregs->_xmm;
+        if (TargetSysHandler != 0 && (xmm_regs[5] & 0xFFFFFFFFFFFFFFFF) != 0x1337133713371337) {
+            xmm_regs[4] = ctx->uc_mcontext.gregs[REG_RAX] & 0xFFFFFFFF;
+            ctx->uc_mcontext.gregs[REG_RAX] = ctx->uc_mcontext.gregs[REG_RCX];
+            ctx->uc_mcontext.gregs[REG_RCX] = TargetSysHandler;
+            ctx->uc_mcontext.gregs[REG_RIP] = TargetSysHandler;
+            return;
+        }
+        if ((xmm_regs[5] & 0xFFFFFFFFFFFFFFFF) == 0x1337133713371337) {
+            xmm_regs[5] = 0;
+            //MESSAGE("SyscallBypassMagic!\n");
+        }
+    }
+
     frame->rip = RIP_sig(sigcontext) + 0xb;
     frame->rcx = RIP_sig(sigcontext);
     frame->eflags = EFL_sig(sigcontext);
@@ -2770,8 +2902,69 @@ static int libc_addr_cb( struct dl_phdr_info *info, size_t size, void *arg )
 }
 #endif
 
+static void patch_kuser_shared_data(void)
+{
+    if (mprotect((void *)0x7FFE0000, 0x1000, PROT_READ | PROT_WRITE) == -1) {
+        MESSAGE("Failed to make kuser_shared_data writable: %s\n", strerror(errno));
+        return;
+    }
+    /* Mark the SystemCall as present. */
+    user_shared_data->SystemCallPad[0] = 1;
+}
+
+static void detect_cpu_vendor(void) {
+    unsigned int eax, ebx, ecx, edx;
+    int avx = 0;
+    if (getenv("PROTON_AVX") != NULL && strcmp(getenv("PROTON_AVX"), "1") == 0) avx = 1;
+
+    __asm__ volatile(
+        "cpuid"
+        : "=a"(eax), "=b"(ebx), "=c"(ecx), "=d"(edx)
+        : "a"(0)
+        : "memory"
+    );
+
+    if (ebx == 0x756E6547 && edx == 0x49656E69 && ecx == 0x6C65746E) {
+        /* GenuineIntel */
+        spoof_leaf1_eax = 0x000A0655;
+        spoof_leaf1_ebx = 0x00200800;
+        if (avx) {
+            spoof_leaf1_ecx = 0x7BFAFBFF;
+        } else spoof_leaf1_ecx = 0x01FAEBFF;
+        spoof_leaf1_edx = 0xBFEBFBFF;
+
+        spoof_leaf40000000_eax = 0x40000001;
+        spoof_leaf40000000_ebx = 0x65707948;
+        spoof_leaf40000000_ecx = 0x67624472;
+        spoof_leaf40000000_edx = 0;
+
+        spoof_leaf40000001_eax = 0x30237648;
+        spoof_leaf40000001_ebx = 0;
+        spoof_leaf40000001_ecx = 0;
+        spoof_leaf40000001_edx = 0;
+    } else if (ebx == 0x68747541 && edx == 0x69746E65 && ecx == 0x444D4163) {
+        /* AuthenticAMD */
+        spoof_leaf1_eax = 0x00A20F12;
+        spoof_leaf1_ebx = 0x00100800;
+        if (avx) {
+            spoof_leaf1_ecx = 0x7AD8320B;
+        } else spoof_leaf1_ecx = 0x00F8220B;
+        spoof_leaf1_edx = 0x178BFBFF;
+
+        spoof_leaf40000000_eax = 0x40000001;
+        spoof_leaf40000000_ebx = 0x706D6953;
+        spoof_leaf40000000_ecx = 0x7653656C;
+        spoof_leaf40000000_edx = 0x2020206D;
+
+        spoof_leaf40000001_eax = 0x30237648;
+        spoof_leaf40000001_ebx = 0;
+        spoof_leaf40000001_ecx = 0;
+        spoof_leaf40000001_edx = 0;
+    }
+}
+
 /**********************************************************************
- *		signal_init_process
+ *	    signal_init_process
  */
 void signal_init_process( TEB *teb )
 {
@@ -2835,6 +3028,15 @@ void signal_init_process( TEB *teb )
     sig_act.sa_sigaction = sigsys_handler;
     if (sigaction( SIGSYS, &sig_act, NULL ) == -1) goto error;
 #endif
+
+#ifdef __linux__
+    denuvo_enabled = getenv("WINE_DENUVO") && !strcmp(getenv("WINE_DENUVO"), "1");
+    if (denuvo_enabled)
+    {
+        detect_cpu_vendor();
+        syscall(SYS_arch_prctl, ARCH_SET_CPUID, 0);
+    }
+#endif
     return;
 
  error:
diff --git a/include/wine/server_protocol.h b/include/wine/server_protocol.h
index efc1cbe..2edee6e 100644
--- a/include/wine/server_protocol.h
+++ b/include/wine/server_protocol.h
@@ -6221,6 +6221,59 @@ struct d3dkmt_mutex_release_reply
 };
 
 
+
+struct set_faketime_request
+{
+    struct request_header __header;
+    char __pad_12[4];
+    unsigned __int64    faketime;
+};
+struct set_faketime_reply
+{
+    struct reply_header __header;
+};
+
+
+
+struct dcomp_create_shared_visual_request
+{
+    struct request_header __header;
+    char __pad_12[4];
+};
+struct dcomp_create_shared_visual_reply
+{
+    struct reply_header __header;
+    obj_handle_t  handle;
+    char __pad_12[4];
+};
+
+
+
+struct dcomp_set_shared_visual_info_request
+{
+    struct request_header __header;
+    obj_handle_t  handle;
+    client_ptr_t  target_root;
+};
+struct dcomp_set_shared_visual_info_reply
+{
+    struct reply_header __header;
+};
+
+
+
+struct dcomp_get_shared_visual_info_request
+{
+    struct request_header __header;
+    obj_handle_t  handle;
+};
+struct dcomp_get_shared_visual_info_reply
+{
+    struct reply_header __header;
+    client_ptr_t  target_root;
+};
+
+
 enum request
 {
     REQ_new_process,
@@ -6530,6 +6583,10 @@ enum request
     REQ_d3dkmt_object_open_name,
     REQ_d3dkmt_mutex_acquire,
     REQ_d3dkmt_mutex_release,
+    REQ_set_faketime,
+    REQ_dcomp_create_shared_visual,
+    REQ_dcomp_set_shared_visual_info,
+    REQ_dcomp_get_shared_visual_info,
     REQ_NB_REQUESTS
 };
 
@@ -6844,6 +6901,10 @@ union generic_request
     struct d3dkmt_object_open_name_request d3dkmt_object_open_name_request;
     struct d3dkmt_mutex_acquire_request d3dkmt_mutex_acquire_request;
     struct d3dkmt_mutex_release_request d3dkmt_mutex_release_request;
+    struct set_faketime_request set_faketime_request;
+    struct dcomp_create_shared_visual_request dcomp_create_shared_visual_request;
+    struct dcomp_set_shared_visual_info_request dcomp_set_shared_visual_info_request;
+    struct dcomp_get_shared_visual_info_request dcomp_get_shared_visual_info_request;
 };
 union generic_reply
 {
@@ -7156,8 +7217,12 @@ union generic_reply
     struct d3dkmt_object_open_name_reply d3dkmt_object_open_name_reply;
     struct d3dkmt_mutex_acquire_reply d3dkmt_mutex_acquire_reply;
     struct d3dkmt_mutex_release_reply d3dkmt_mutex_release_reply;
+    struct set_faketime_reply set_faketime_reply;
+    struct dcomp_create_shared_visual_reply dcomp_create_shared_visual_reply;
+    struct dcomp_set_shared_visual_info_reply dcomp_set_shared_visual_info_reply;
+    struct dcomp_get_shared_visual_info_reply dcomp_get_shared_visual_info_reply;
 };
 
-#define SERVER_PROTOCOL_VERSION 956
+#define SERVER_PROTOCOL_VERSION 957
 
 #endif /* __WINE_WINE_SERVER_PROTOCOL_H */
diff --git a/server/d3dkmt.c b/server/d3dkmt.c
index b066021..dde35ed 100644
--- a/server/d3dkmt.c
+++ b/server/d3dkmt.c
@@ -876,6 +876,13 @@ static struct composition *create_composition(void)
     return composition;
 }
 
+/* Set faketime for Denuvo */
+unsigned long long denuvo_faketime_enabled = 0;
+DECL_HANDLER(set_faketime)
+{
+    denuvo_faketime_enabled = req->faketime;
+}
+
 /* Create a DirectComposition shared visual handle */
 DECL_HANDLER(dcomp_create_shared_visual)
 {
diff --git a/server/protocol.def b/server/protocol.def
index 5f6e6c0..2de91b1 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -4325,6 +4325,12 @@ enum inproc_sync_type
 @END
 
 
+/* Set faketime for Denuvo */
+@REQ(set_faketime)
+    unsigned __int64    faketime;
+@END
+
+
 /* Create a DirectComposition shared visual handle */
 @REQ(dcomp_create_shared_visual)
 @REPLY
diff --git a/server/request_handlers.h b/server/request_handlers.h
index 8279782..1422a92 100644
--- a/server/request_handlers.h
+++ b/server/request_handlers.h
@@ -314,6 +314,10 @@ DECL_HANDLER(d3dkmt_share_objects);
 DECL_HANDLER(d3dkmt_object_open_name);
 DECL_HANDLER(d3dkmt_mutex_acquire);
 DECL_HANDLER(d3dkmt_mutex_release);
+DECL_HANDLER(set_faketime);
+DECL_HANDLER(dcomp_create_shared_visual);
+DECL_HANDLER(dcomp_set_shared_visual_info);
+DECL_HANDLER(dcomp_get_shared_visual_info);
 
 typedef void (*req_handler)( const void *req, void *reply );
 static const req_handler req_handlers[REQ_NB_REQUESTS] =
@@ -625,6 +629,10 @@ static const req_handler req_handlers[REQ_NB_REQUESTS] =
     (req_handler)req_d3dkmt_object_open_name,
     (req_handler)req_d3dkmt_mutex_acquire,
     (req_handler)req_d3dkmt_mutex_release,
+    (req_handler)req_set_faketime,
+    (req_handler)req_dcomp_create_shared_visual,
+    (req_handler)req_dcomp_set_shared_visual_info,
+    (req_handler)req_dcomp_get_shared_visual_info,
 };
 
 C_ASSERT( sizeof(abstime_t) == 8 );
@@ -2385,3 +2393,15 @@ C_ASSERT( offsetof(struct d3dkmt_mutex_release_request, key_value) == 20 );
 C_ASSERT( offsetof(struct d3dkmt_mutex_release_request, fence_value) == 24 );
 C_ASSERT( offsetof(struct d3dkmt_mutex_release_request, runtime_size) == 32 );
 C_ASSERT( sizeof(struct d3dkmt_mutex_release_request) == 40 );
+C_ASSERT( offsetof(struct set_faketime_request, faketime) == 16 );
+C_ASSERT( sizeof(struct set_faketime_request) == 24 );
+C_ASSERT( sizeof(struct dcomp_create_shared_visual_request) == 16 );
+C_ASSERT( offsetof(struct dcomp_create_shared_visual_reply, handle) == 8 );
+C_ASSERT( sizeof(struct dcomp_create_shared_visual_reply) == 16 );
+C_ASSERT( offsetof(struct dcomp_set_shared_visual_info_request, handle) == 12 );
+C_ASSERT( offsetof(struct dcomp_set_shared_visual_info_request, target_root) == 16 );
+C_ASSERT( sizeof(struct dcomp_set_shared_visual_info_request) == 24 );
+C_ASSERT( offsetof(struct dcomp_get_shared_visual_info_request, handle) == 12 );
+C_ASSERT( sizeof(struct dcomp_get_shared_visual_info_request) == 16 );
+C_ASSERT( offsetof(struct dcomp_get_shared_visual_info_reply, target_root) == 8 );
+C_ASSERT( sizeof(struct dcomp_get_shared_visual_info_reply) == 16 );
diff --git a/server/request_trace.h b/server/request_trace.h
index 497c88e..5d29d02 100644
--- a/server/request_trace.h
+++ b/server/request_trace.h
@@ -3524,6 +3524,36 @@ static void dump_d3dkmt_mutex_release_request( const struct d3dkmt_mutex_release
     dump_varargs_bytes( ", runtime=", cur_size );
 }
 
+static void dump_set_faketime_request( const struct set_faketime_request *req )
+{
+    dump_uint64( " faketime=", &req->faketime );
+}
+
+static void dump_dcomp_create_shared_visual_request( const struct dcomp_create_shared_visual_request *req )
+{
+}
+
+static void dump_dcomp_create_shared_visual_reply( const struct dcomp_create_shared_visual_reply *req )
+{
+    fprintf( stderr, " handle=%04x", req->handle );
+}
+
+static void dump_dcomp_set_shared_visual_info_request( const struct dcomp_set_shared_visual_info_request *req )
+{
+    fprintf( stderr, " handle=%04x", req->handle );
+    dump_uint64( ", target_root=", &req->target_root );
+}
+
+static void dump_dcomp_get_shared_visual_info_request( const struct dcomp_get_shared_visual_info_request *req )
+{
+    fprintf( stderr, " handle=%04x", req->handle );
+}
+
+static void dump_dcomp_get_shared_visual_info_reply( const struct dcomp_get_shared_visual_info_reply *req )
+{
+    dump_uint64( " target_root=", &req->target_root );
+}
+
 typedef void (*dump_func)( const void *req );
 
 static const dump_func req_dumpers[REQ_NB_REQUESTS] =
@@ -3835,6 +3865,10 @@ static const dump_func req_dumpers[REQ_NB_REQUESTS] =
     (dump_func)dump_d3dkmt_object_open_name_request,
     (dump_func)dump_d3dkmt_mutex_acquire_request,
     (dump_func)dump_d3dkmt_mutex_release_request,
+    (dump_func)dump_set_faketime_request,
+    (dump_func)dump_dcomp_create_shared_visual_request,
+    (dump_func)dump_dcomp_set_shared_visual_info_request,
+    (dump_func)dump_dcomp_get_shared_visual_info_request,
 };
 
 static const dump_func reply_dumpers[REQ_NB_REQUESTS] =
@@ -4146,6 +4180,10 @@ static const dump_func reply_dumpers[REQ_NB_REQUESTS] =
     (dump_func)dump_d3dkmt_object_open_name_reply,
     (dump_func)dump_d3dkmt_mutex_acquire_reply,
     NULL,
+    NULL,
+    (dump_func)dump_dcomp_create_shared_visual_reply,
+    NULL,
+    (dump_func)dump_dcomp_get_shared_visual_info_reply,
 };
 
 static const char * const req_names[REQ_NB_REQUESTS] =
@@ -4457,6 +4495,10 @@ static const char * const req_names[REQ_NB_REQUESTS] =
     "d3dkmt_object_open_name",
     "d3dkmt_mutex_acquire",
     "d3dkmt_mutex_release",
+    "set_faketime",
+    "dcomp_create_shared_visual",
+    "dcomp_set_shared_visual_info",
+    "dcomp_get_shared_visual_info",
 };
 
 static const struct
